### PR TITLE
Add mechanism for explicit marking of fixtures which should be run with asyncio

### LINKED ISSUE #124: Require explicit marks for async fixtures
`pytest-asyncio` interprets all `async def` based fixtures as things that should be run by the `asyncio` event loop.  This appears to make it impossible for a test suite to have side-by-side tests for both libraries.

```python
import asyncio
import trio

import pytest

import pytest_trio


@pytest_trio.trio_fixture
async def trio_thing():
    await trio.sleep(0)
    return 'trio'


@pytest.fixture
async def asyncio_thing():
    await asyncio.sleep(0)
    return 'asyncio'


@pytest.mark.asyncio
async def test_something_asyncio(asyncio_thing):
    await asyncio.sleep(0)
    assert asyncio_thing == 'asyncio'


@pytest.mark.trio
async def test_something_trio(trio_thing):
    await trio.sleep(0)
    assert trio_thing == 'trio'
```

This results in:

```shell
$ pytest test_it.py -s
=========================================================================================== test session starts ===========================================================================================
platform linux -- Python 3.6.5, pytest-4.0.2, py-1.8.0, pluggy-0.9.0 -- /home/piper/python-environments/lahja/bin/python
cachedir: .pytest_cache
rootdir: /home/piper/projects/lahja, inifile: pytest.ini
plugins: xdist-1.25.0, trio-0.5.2, forked-1.0.2, asyncio-0.10.0
collected 2 items

test_it.py::test_something_asyncio PASSED
test_it.py::test_something_trio ERROR

================================================================================================= ERRORS ==================================================================================================
__________________________________________________________________________________ ERROR at setup of test_something_trio __________________________________________________________________________________

    @pytest_trio.trio_fixture
    async def trio_thing():
>       await trio.sleep(0)


test_it.py:11:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../../python-environments/lahja/lib/python3.6/site-packages/trio/_timeouts.py:83: in sleep
    await _core.checkpoint()
../../../python-environments/lahja/lib/python3.6/site-packages/trio/_core/_run.py:1697: in checkpoint
    with CancelScope(deadline=-inf):
../../../python-environments/lahja/lib/python3.6/site-packages/trio/_core/_ki.py:165: in wrapper
    return fn(*args, **kwargs)
../../../python-environments/lahja/lib/python3.6/site-packages/trio/_core/_run.py:179: in __enter__
    task = _core.current_task()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    def current_task():
        """Return the :class:`Task` object representing the current task.

        Returns:
          Task: the :class:`Task` that called :func:`current_task`.

        """

        try:
            return GLOBAL_RUN_CONTEXT.task
        except AttributeError:
>           raise RuntimeError("must be called from async context") from None
E           RuntimeError: must be called from async context


../../../python-environments/lahja/lib/python3.6/site-packages/trio/_core/_run.py:1645: RuntimeError
```

### PR DESCRIPTION
**Pull request by pipermerriam**
- Fixes #124 

Opening this up to get feedback from the maintainers.

With this patch existing users of `pytest-asyncio` should not experience any change in behavior.

This exposes a new API `pytest_asyncio.asyncio_fixture` that is intended to be used like this:

```python
import pytest_asyncio


@pytest_asyncio.asyncio_fixture
async def server():   # async yield fixture
    await server.start()
    yield
    await server.stop()

@pytest_asyncio.asyncio_fixture
async def client():  # normal async function
    await client.connect_to_server()
    return client
```

This PR also adds a `pytest.ini` configuration value: `asyncio_mode` which defaults to `True`.  When set to `False`, **only** fixtures with the above mark (as well as a few other conditions) will be run by the asyncio event loop.

Some notes:

- I did not see a test suite.  This was tested manually against a local project.  If you'd like me to write up some tests I don't mind, but it felt heavy handed to do that without the blessing of a maintainer.
- Looking for feedback on whether this is a feature that will be accepted before I polish this and make it nice.

### CODE REVIEW

**Comment by pipermerriam** [comment]
I present an offering of cute animal picture for a moment of time from the maintainer gods.  :pray: 

e7c924f471fdf1699f44547a002ec02c--mini-goats-goats-baby

**Comment by asvetlov** [comment]
I'm not a maintainer but have 5 cents to add.

I think we need not a bool flag but three-state:
1. "on": automatically apply `pytest.mark.asyncio` to all discovered asynchronous fixtures and tests
2. "off": do nothing with autoapplying
3. "compat" (which is default and deprecated): the current behavior. Apply asyncio mark to async fixtures but don't apply to async tests.

Having default "compat" mode deprecated we notice people switch either to "on" or "off" in their pytest configuration to avoid deprecation warning displayed by pytest.
After some deprecation period we can switch default and even drop "compat" mode eventually.

**Comment by asvetlov** [comment]
@pipermerriam do you want to implement `on`/`off`/`compat` mode?

**Comment by pipermerriam** [comment]
@asvetlov yes, but it will be about 2 weeks before I pick this back up.  Life and work things taking front seat for a bit and won't be able to focus on this for till then.  (won't be offended if anyone takes what I started and moves forward with it)

**Comment by asvetlov** [comment]
I'm personally is overwhelmed as well by my job and fixing CPython 3.8 beta.
I had a hope to finish 3.8 hacking this week but seems I need more time.
There is a low chance that I can pick up your job in a week or two.

**Comment by graingert** [comment]
@pipermerriam I had a go at resolving the conflicts, but I'm not sure what needs doing in pytest_asyncio/plugin.py

Gentoo are currently having trouble when running the anyio test suite with pytest-asyncio installed

**Comment by graingert** [comment]
@asvetlov thanks for taking this on!

File: README.rst [hunk 1]
```diff
@@ -51,6 +54,70 @@ To install pytest-asyncio, simply:
 
 This is enough for pytest to pick up pytest-asyncio.
 
+Modes
+-----
+
+Strting from ``pytest-asyncio>=0.17``, three modes are provided: *auto*, *strict* and
```
Comment (graingert) [inline README.rst:60]:
```suggestion
Starting from ``pytest-asyncio>=0.17``, three modes are provided: *auto*, *strict* and
```

**Comment by asvetlov** [comment]
The PR is done and tested.

@Tinche @seifertm I very much appreciate your review.

File: README.rst (hunk 1 above)
Comment (asvetlov) [reply README.rst]: Thanks!

File: README.rst [hunk 2]
```diff
@@ -51,6 +54,70 @@ To install pytest-asyncio, simply:
 
 This is enough for pytest to pick up pytest-asyncio.
 
+Modes
+-----
+
+Starting from ``pytest-asyncio>=0.17``, three modes are provided: *auto*, *strict* and
+*legacy* (deault).
```
Comment (Tinche) [inline README.rst:61]:
```suggestion
*legacy* (default).
```

File: README.rst [hunk 3]
```diff
@@ -51,6 +54,70 @@ To install pytest-asyncio, simply:
...
+if they have no ``@pytest.mark.asyncio`` marker.
+
+All async fixtures are considered *asyncio-driven* as well, even if they are decorated
+with a regular ``@pytest.fixture`` decorator instead of dedicated
+``@pytest_asyncio.fixture`` counterpart.
+
+*asyncio-driven* means that tests and fixtures are executed by ``pytest-asyncio``
+plugin.
+
+This mode requires the simpliest tests and fixtures configuration and is
```
Comment (Tinche) [inline README.rst:91]:
```suggestion
This mode requires the simplest tests and fixtures configuration and is
```

**Review by Tinche — APPROVED** [review]
Left suggestions for a couple of typos, looks good overall. A really good change!

**Review by seifertm — APPROVED** [review]
This is great! Tests shouldn't just break when pytest-asyncio is installed in addition to any other package. The new modes are also more consistent than the current behavior.

 I like that we switched from `@pytest_asyncio.asyncio_fixture` to the shorter `@pytest_asyncio.fixture`. I also appreciate that `asyncio_mode=strict` is the new default.

I annotated a few typos and issues in the docs. Other than that: Great job!

**Comment by asvetlov** [comment]
Thanks, guys!

**Pull request merged**